perun/logic/temp.py

```diff
@@ -0,0 +1,653 @@
+
+import perun.logic.pcs as pcs
+import perun.logic.store as store
+import perun.utils.exceptions as exceptions
+
+
+UNPROTECTED, PROTECTED = 'unprotected', 'protected'
+
+SORT_ATTR = [
+    'name', 'protection', 'size'
+]
+PROTECTION_LEVEL = [
+    'all', UNPROTECTED, PROTECTED
+]
+
+SORT_ATTR_MAP = {
+    'name': {
+        'pos': 0,
+        'reverse': False
+    },
+    'protection': {
+        'pos': 1,
+        'reverse': False
+    },
+    'size': {
+        'pos': 2,
+        'reverse': True
+    }
+}
+
+
+def temp_path(path):
+    """Transforms the provided path to the tmp/ directory context, i.e.:
+        - absolute path located in the .perun/tmp/ directory: no change
+        - relative path: transformed to absolute path where .perun/tmp/ is considered as the
+                         root/working directory for the relative path
+        - absolute or relative path out of the .perun/tmp/: an exception is thrown
+
+    :param str path: the path that should be transformed
+
+    :return str: the transformed path
+    """
+    return _join_to_tmp_path(path)
+
+
+def touch_temp_dir(dir_path):
+    """Touches the given directory path.
+
+    For details regarding the path format, see the module docstring.
+
+    :param str dir_path: the directory path
+    """
+    # Append the path to the tmp/ directory
+    dir_path = _join_to_tmp_path(dir_path)
+    # Create the new directory(ies)
+    if not os.path.exists(dir_path):
+        os.makedirs(dir_path)
+
+
+def touch_temp_file(file_path, protect=False):
+    """Touches the given temporary file and sets the protection level.
+
+    For details regarding the path format, see the module docstring.
+
+    :param str file_path: the temporary file path
+    :param bool protect: if True, the temporary file will be indexed as protected
+    """
+    # Append the file path to the tmp/ directory
+    file_path = _join_to_tmp_path(file_path.rstrip(os.sep))
+    # Make sure that the directory hierarchy for the file exists
+    touch_temp_dir(os.path.dirname(file_path))
+
+    store.touch_file(file_path)
+    # Register the file as protected if needed
+    if protect:
+        _add_to_index(file_path, protected=protect)
+
+
+def exists_temp_dir(dir_path):
+    """Checks if the temporary directory given by 'dir_path' exists.
+
+    For details regarding the path format, see the module docstring.
+
+    :param str dir_path: the temporary directory
+
+    :return bool: True if the directory exists
+    """
+    dir_path = _join_to_tmp_path(dir_path)
+    return os.path.exists(dir_path) and os.path.isdir(dir_path)
+
+
+def exists_temp_file(file_path):
+    """Checks if the temporary file given by 'file_path' exists.
+
+    For details regarding the path format, see the module docstring.
+
+    :param str file_path: the temporary file
+
+    :return bool: True if the file exists
+    """
+    file_path = _join_to_tmp_path(file_path)
+    return os.path.exists(file_path) and os.path.isfile(file_path)
```

Comment:
Note: While I see how this could be refactored (extract os.path._ functions to parameter), I am not sure if it is worth it...

Reply:
I agree, it would either create one more useless call layer in case of internal private function or complicate the use of the function for the users.